internal/provider/repository_branch_protection_resource.go

```diff
@@ -0,0 +1,816 @@
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*forgejo.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf(
+				"Expected *forgejo.Client, got: %T. Please report this issue to the provider developers.",
+				req.ProviderData,
+			),
+		)
+		return
+	}
+
+	r.client = client
+}
+
+func (r *repositoryBranchProtectionResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	defer un(trace(ctx, "Create repository branch protection resource"))
+
+	var data repositoryBranchProtectionResourceModel
+
+	diags := req.Plan.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Info(ctx, "Create repository branch protection", map[string]any{
+		"owner":       data.Owner.ValueString(),
+		"repo":        data.Repo.ValueString(),
+		"branch_name": data.BranchName.ValueString(),
+	})
+
+	// Convert model to API request
+	opts := r.modelToCreateOption(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create branch protection
+	_, res, err := r.client.CreateBranchProtection(
+		data.Owner.ValueString(),
+		data.Repo.ValueString(),
+		opts,
+	)
+	if err != nil {
+		tflog.Error(ctx, "Error", map[string]any{
+			"error": err.Error(),
+		})
+
+		var msg string
+		if res != nil {
+			switch res.StatusCode {
+			case 403:
+				msg = fmt.Sprintf("Forbidden: %s", err)
+			case 404:
+				msg = fmt.Sprintf("Repository not found: %s", err)
+			case 422:
+				msg = fmt.Sprintf("Validation error: %s", err)
+			default:
+				msg = fmt.Sprintf("Unknown error: %s", err)
+			}
+		} else {
+			msg = fmt.Sprintf("Failed to create branch protection: %s", err)
+		}
+		resp.Diagnostics.AddError("Unable to create branch protection", msg)
+		return
+	}
+
+	// Set the ID
+	data.ID = types.StringValue(fmt.Sprintf("%s/%s/%s",
+		data.Owner.ValueString(),
+		data.Repo.ValueString(),
+		data.BranchName.ValueString()))
+
+	// Update model with response data
+	tflog.Trace(ctx, "created branch protection resource")
+
+	diags = resp.State.Set(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r *repositoryBranchProtectionResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	defer un(trace(ctx, "Read repository branch protection resource"))
+
+	var data repositoryBranchProtectionResourceModel
+
+	diags := req.State.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get branch protection
+	_, res, err := r.client.GetBranchProtection(
+		data.Owner.ValueString(),
+		data.Repo.ValueString(),
+		data.BranchName.ValueString(),
+	)
+	if err != nil {
+		if res != nil && res.StatusCode == 404 {
+			// Branch protection no longer exists
+			tflog.Trace(ctx, "branch protection not found, removing from state")
+			resp.State.RemoveResource(ctx)
```

Comment:
Please don't do this. If the resource was deleted outside of Terraform then I'd consider this an error, and the provider should fail. Don't siltently remove the resource from state - I consider this dangerous.

Please check e.g. https://github.com/svalabs/terraform-provider-forgejo/blob/edff37edbd29028fad2f3a4a2b00a7da5fcfc37a/internal/provider/user_resource.go#L533 for an example of handling 404 errors.